src/lsm/grid.zig

```diff
@@ -78,11 +166,72 @@ pub fn GridType(comptime Storage: type) type {
         ) void {
             assert(grid.superblock.opened);
             assert(address > 0);
-            // TODO Assert that address is acquired in the free set.
-            // TODO Assert that the block ptr is not being used for another I/O (read or write).
-            // TODO Assert that block is not already writing.
+            assert(!grid.superblock.free_set.is_free(address));
 
-            grid.superblock.storage.write_sectors(callback, write, block, block_offset(address));
+            // Assert that block is not already writing.
+            // Assert that the block ptr is not being used for another I/O (read or write).
+            {
+                var it = grid.writes_pending.peek();
+                while (it) |pending_write| : (it = pending_write.next) {
+                    assert(address != pending_write.address);
+                    assert(block != pending_write.block);
+                }
+            }
+            {
+                var it = grid.write_iops.iterate();
+                while (it.next()) |iop| {
+                    assert(address != iop.write.address);
+                    assert(block != iop.write.block);
+                }
+            }
+            {
+                var it = grid.read_iops.iterate();
+                while (it.next()) |iop| {
+                    assert(address != iop.reads.peek().?.address);
```

Comment:
> This means that a write also can't start until any reads for the same `address` complete or it would trip this assertion. However, this may be fine, since we won't overlap prefetching or reads at the same time as compaction (which is the only place we write to blocks) in the commit log ratchet mechanism.

Yes, I realize that there may be some use-cases for this API where racing writes and reads aren't problematic, but as you say our use-case doesn't need to read and write the same address at the same time. If we find a use-case where we do need to support that it will be easy to remove this check.

> It would be good to do these asserts rather at the moment that we start the I/O, so that there's no gap from when we do the assertions and when we issue the I/O. This will then also match `start_read()`.

I can see the argument for that, but I think it would be even stronger to have the assertions at both places :+1:

Reply:
😄 I was actually going to suggest checking in both places. What was your thinking on why it didn't make sense?

Reply:
I think I said that because I saw putting in both places as superfluous and came around to seeing `start_write()` as the obvious/ideal place to put it.

I don't see any reason to not have both though, so I've gone ahead and done that. I also added some missing parallel assertions on the read path ee5db252b5ccf245d5cae742fb19850f4d076a83